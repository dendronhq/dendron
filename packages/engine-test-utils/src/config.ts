import { DConfig, readYAML, writeYAML } from "@dendronhq/common-server";
import { IntermediateDendronConfig } from "@dendronhq/common-all";

export class TestConfigUtils {
  static getConfig = (opts: { wsRoot: string }) => {
    const configPath = DConfig.configPath(opts.wsRoot);
    const config = readYAML(configPath) as IntermediateDendronConfig;
    return config;
  };

  static withConfig = (
    func: (config: IntermediateDendronConfig) => IntermediateDendronConfig,
    opts: { wsRoot: string }
  ) => {
    const config = TestConfigUtils.getConfig(opts);

    const newConfig = func(config);
    TestConfigUtils.writeConfig({ config: newConfig, wsRoot: opts.wsRoot });
    return newConfig;
  };

  static writeConfig = (opts: {
    config: IntermediateDendronConfig;
    wsRoot: string;
  }) => {
    const configPath = DConfig.configPath(opts.wsRoot);
    return writeYAML(configPath, opts.config);
  };
}
