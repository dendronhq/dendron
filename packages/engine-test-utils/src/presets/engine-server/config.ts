import { CONSTANTS, ConfigUtils } from "@dendronhq/common-all";
import { DConfig, readYAML, writeYAML } from "@dendronhq/common-server";
import { TestPresetEntryV4 } from "@dendronhq/common-test-utils";
import path from "path";

function genDefaultConfig() {
  return ConfigUtils.genDefaultConfig();
}

const WRITE = {
  NEW_CONFIG: new TestPresetEntryV4(async ({ wsRoot }) => {
    const config = genDefaultConfig();
    ConfigUtils.setPublishProp(config, "copyAssets", false);
    await DConfig.writeConfig({ wsRoot, config });
    const cpath = path.join(wsRoot, CONSTANTS.DENDRON_CONFIG_FILE);
    const configOnFile = readYAML(cpath);
    return [
      {
        actual: configOnFile,
        expected: config,
      },
    ];
  }),
};

const GET = {
  DEFAULT_CONFIG: new TestPresetEntryV4(async ({ wsRoot }) => {
    const cpath = path.join(wsRoot, CONSTANTS.DENDRON_CONFIG_FILE);
    writeYAML(cpath, genDefaultConfig());
    const config = DConfig.readConfigSync(wsRoot);
    return [
      {
        actual: config,
        expected: genDefaultConfig(),
      },
    ];
  }),
};

export const ENGINE_CONFIG_PRESETS = {
  WRITE,
  GET,
};
