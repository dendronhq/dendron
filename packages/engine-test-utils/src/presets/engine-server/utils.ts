import { NoteProps, SchemaUtils } from "@dendronhq/common-all";
import {
  CreateNoteFactory,
  NOTE_PRESETS_V4,
  NoteTestUtilsV4,
  PreSetupHookFunction,
  SCHEMA_PRESETS_V4,
} from "@dendronhq/common-test-utils";
import _ from "lodash";
import { resolvePath } from "@dendronhq/common-server";
import path from "path";
import * as fs from "fs";

/**
 * Notes created:
 *   - vault1:
 *     - foo
 *     - foo.ch1
 *   - vault2:
 *     - bar
 */
export const setupBasic: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
  extra,
}) => {
  const vault = vaults[0];
  // TODO: HACK
  let props;
  if (extra?.idv2) {
    props = { id: "foo-id" };
  }
  await NOTE_PRESETS_V4.NOTE_SIMPLE.create({
    vault,
    wsRoot,
    props,
  });
  await NOTE_PRESETS_V4.NOTE_SIMPLE_CHILD.create({
    vault,
    wsRoot,
  });
  await NOTE_PRESETS_V4.NOTE_SIMPLE_OTHER.create({
    vault,
    wsRoot,
  });
  await SCHEMA_PRESETS_V4.SCHEMA_SIMPLE.create({ vault, wsRoot });
};

/**
 <pre>
 /vault1
 ├── bar.ch1.gch1.ggch1.md
 ├── bar.ch1.gch1.md
 ├── bar.ch1.md
 ├── bar.md
 ├── foo.ch1.gch1.ggch1.md
 ├── foo.ch1.gch1.md
 ├── foo.ch1.gch2.md
 ├── foo.ch1.md
 ├── foo.ch2.md
 ├── foo.md
 ├── root.md
 └── root.schema.yml

/vault2
 ├── root.md
 └── root.schema.yml

 /vault3
 ├── root.md
 └── root.schema.yml
</pre>
 * */
export const setupHierarchyForLookupTests: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  const opts = {
    vault: vaults[0],
    wsRoot,
  };
  const fnames = [
    "foo",
    "foo.ch1",
    "foo.ch1.gch1",
    "foo.ch1.gch1.ggch1",
    "foo.ch1.gch2",
    "foo.ch2",
    "bar",
    "bar.ch1",
    "bar.ch1.gch1",
    "bar.ch1.gch1.ggch1",
    "goo.ends-with-ch1.no-ch1-by-itself",
  ];

  for (const fname of fnames) {
    await CreateNoteFactory({ fname, body: `${fname} body` }).create(opts);
  }
};

export const setupJournals: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  const vault = vaults[0];
  const names = [
    "daily",
    "daily.journal",
    "daily.journal.2020",
    "daily.journal.2020.07",
    "daily.journal.2020.07.01.one",
    "daily.journal.2020.07.05.two",
  ];
  return Promise.all(
    names.map((fname) => {
      return NoteTestUtilsV4.createNote({
        wsRoot,
        vault,
        fname,
      });
    })
  );
};

// Workspace will look like:
// .
// ├── dendron.code-workspace
// ├── dendron.yml
// ├── vault1
// │   ├── foo.ch1.md
// │   ├── foo.md
// │   ├── foo.schema.yml
// │   ├── root.md
// │   └── root.schema.yml
// ├── vault2
// │   ├── bar.md
// │   ├── root.md
// │   └── root.schema.yml
// └── vault3
//     ├── root.md
//     └── root.schema.yml
export const setupBasicMulti: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  const vault1 = _.find(vaults, { fsPath: "vault1" })!;
  const vault2 = _.find(vaults, { fsPath: "vault2" })!;
  await NOTE_PRESETS_V4.NOTE_SIMPLE.create({
    vault: vault1,
    wsRoot,
  });
  await NOTE_PRESETS_V4.NOTE_SIMPLE_CHILD.create({
    vault: vault1,
    wsRoot,
  });
  await NOTE_PRESETS_V4.NOTE_SIMPLE_OTHER.create({
    vault: vault2,
    wsRoot,
  });
  await SCHEMA_PRESETS_V4.SCHEMA_SIMPLE.create({ vault: vault1, wsRoot });
};

/**
 *
 * - foo
 * ```
 * ![[foo.one]]
 * ```
 *
 * - foo.one
 * ```
 * # Foo.One
 * ![[foo.two]]
 * Regular wikilink: [[foo.two]]
 * ```
 *
 * - foo.two
 * ```
 * # Foo.Two
 * blah
 * ```
 *
 *
 */
export const setupNoteRefRecursive: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
  extra,
}) => {
  const vault = vaults[0];
  // TODO: HACK
  let props;
  if (extra?.idv2) {
    props = { id: "foo-id" };
  }
  await NOTE_PRESETS_V4.NOTE_SIMPLE.create({
    vault,
    wsRoot,
    body: "![[foo.one]]",
    props,
  });
  await NoteTestUtilsV4.createNote({
    vault,
    fname: "foo.one",
    body: ["# Foo.One", `![[foo.two]]`, `Regular wikilink: [[foo.two]]`].join(
      "\n"
    ),
    wsRoot,
    props: {
      id: "foo.one-id",
    },
  });
  await NoteTestUtilsV4.createNote({
    vault,
    fname: "foo.two",
    body: ["# Foo.Two", `blah`].join("\n"),
    wsRoot,
  });
};

/**
 * Diamond schema is laid out such that:
 *    bar
 *   /   \
 * ch1   ch2
 *   \   /
 *    gch
 * */
export const setupSchemaWithDiamondGrandchildren: PreSetupHookFunction = async (
  opts
) => {
  await setupBasic(opts);
  const { wsRoot, vaults } = opts;
  const vault = vaults[0];
  NoteTestUtilsV4.createSchema({
    fname: "bar",
    wsRoot,
    vault,
    modifier: (schema) => {
      const schemas = [
        SchemaUtils.createFromSchemaOpts({
          id: "bar",
          parent: "root",
          fname: "bar",
          children: ["ch1", "ch2"],
          vault,
        }),
        SchemaUtils.createFromSchemaRaw({
          id: "ch1",
          children: ["gch"],
          vault,
        }),
        SchemaUtils.createFromSchemaRaw({
          id: "ch2",
          children: ["gch"],
          vault,
        }),
        SchemaUtils.createFromSchemaRaw({
          id: "gch",
          template: { id: "template.gch", type: "note" },
          vault,
        }),
      ];
      schemas.map((s) => {
        schema.schemas[s.id] = s;
      });
      return schema;
    },
  });
  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "gch template",
    fname: "template.gch",
    vault,
  });
};

export const setupSchemaPreseet: PreSetupHookFunction = async (opts) => {
  await setupBasic(opts);
  const { wsRoot, vaults } = opts;
  const vault = vaults[0];
  NoteTestUtilsV4.createSchema({
    fname: "bar",
    wsRoot,
    vault,
    modifier: (schema) => {
      const schemas = [
        SchemaUtils.createFromSchemaOpts({
          id: "bar",
          parent: "root",
          fname: "bar",
          children: ["ch1", "ch2"],
          vault,
        }),
        SchemaUtils.createFromSchemaRaw({
          id: "ch1",
          template: { id: "bar.template.ch1", type: "note" },
          vault,
        }),
        SchemaUtils.createFromSchemaRaw({
          id: "ch2",
          template: { id: "bar.template.ch2", type: "note" },
          namespace: true,
          vault,
        }),
      ];
      schemas.map((s) => {
        schema.schemas[s.id] = s;
      });
      return schema;
    },
  });
  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "ch1 template",
    fname: "bar.template.ch1",
    props: { tags: "tag-foo" },
    vault,
  });
  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "ch2 template",
    fname: "bar.template.ch2",
    vault,
  });
};

/**
 * Diamond schema is laid out such that:
 *    bar
 *   /   \
 * ch1   ch2 (namespace: true)
 *   \   /
 *    gch
 * */
export const setupSchemaWithDiamondAndParentNamespace: PreSetupHookFunction =
  async (opts) => {
    await setupBasic(opts);
    const { wsRoot, vaults } = opts;
    const vault1 = vaults[0];

    const withDiamond = path.join(
      resolvePath(vault1.fsPath, wsRoot),
      "withDiamond.schema.yml"
    );
    fs.writeFileSync(
      withDiamond,
      `
version: 1
schemas:
  - id: withDiamond
    children:
      - ch1
      - ch2
    title: withDiamond
    parent: root
  - id: ch1
    children:
      - gch
  - id: ch2
    namespace: true
    children:
      - gch
  - id: gch
    template: template.test
`
    );

    await NoteTestUtilsV4.createNote({
      wsRoot,
      body: "Template text",
      fname: "template.test",
      vault: vault1,
    });
  };

/**
 * Sets up schema which includes a schema that has Diamond grandchildren
 *
 * */
export const setupSchemaWithIncludeOfDiamond: PreSetupHookFunction = async (
  opts
) => {
  await setupBasic(opts);
  const { wsRoot, vaults } = opts;
  const vault1 = vaults[0];

  const withDiamond = path.join(
    resolvePath(vault1.fsPath, wsRoot),
    "withDiamond.schema.yml"
  );
  fs.writeFileSync(
    withDiamond,
    `
version: 1
schemas:
  - id: withDiamond
    children:
      - ch1
      - ch2
    title: withDiamond
    parent: root
  - id: ch1
    children:
      - gch
  - id: ch2
    children:
      - gch
  - id: gch
    template: template.test
`
  );

  const includesDiamondPath = path.join(
    resolvePath(vault1.fsPath, wsRoot),
    "includesDiamond.schema.yml"
  );
  fs.writeFileSync(
    includesDiamondPath,
    `
version: 1

imports:
  - withDiamond

schemas:
  - id: includesDiamond
    parent: root
    namespace: true
    children:
      - a-ch1
      - a-ch2
  - id: a-ch1
    children:
      - withDiamond.gch
  - id: a-ch2
    children:
      - withDiamond.gch
`
  );

  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "Template text",
    fname: "template.test",
    vault: vault1,
  });
};

/**
 * Sets up workspace which has a schema that uses YAML expansion syntax ('<<' type of expansion). */
export const setupSchemaWithExpansion: PreSetupHookFunction = async (opts) => {
  await setupBasic(opts);
  const { wsRoot, vaults } = opts;
  const vault1 = vaults[0];

  const withExpansion = path.join(
    resolvePath(vault1.fsPath, wsRoot),
    "withExpansion.schema.yml"
  );
  fs.writeFileSync(
    withExpansion,
    `
_anchors:
  projects: &projects
    title: projects
    parent: root
    template: templates.projects
    
version: 1
imports: []

schemas:
  - <<: *projects
    id: proj
  - <<: *projects
    id: op
`
  );

  const inlineSchemaPath = path.join(
    resolvePath(vault1.fsPath, wsRoot),
    "includesExpansion.schema.yml"
  );
  fs.writeFileSync(
    inlineSchemaPath,
    `
version: 1

imports:
  - withExpansion

schemas:
  - id: includer
    parent: root
    namespace: true
    children:
      - withExpansion.proj
`
  );

  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "Template text",
    fname: "templates.projects",
    vault: vault1,
  });
};

export const setupInlineSchema: PreSetupHookFunction = async (opts) => {
  await setupBasic(opts);
  const { wsRoot, vaults } = opts;
  const vault1 = vaults[0];

  const inlineSchemaPath = path.join(
    resolvePath(vault1.fsPath, wsRoot),
    "inlined.schema.yml"
  );
  fs.writeFileSync(
    inlineSchemaPath,
    `
version: 1
imports: 
  - foo
schemas:
  - id: plain_schema
    parent: root
    children:
      - plain_schema_child
      - daily
  - id: daily
    parent: root
    children:
      - id: journal
        children:
          - pattern: "[0-2][0-9][0-9][0-9]"
            title: year
            id: year_id
            children:
              - pattern: "[0-1][0-9]"
                children:
                  - pattern: "[0-3][0-9]"
                    title: day
                    template:
                      id: templates.day
                      type: note
  - id: id_with_imported_child
    children:
      - foo.foo
  - id: with_child_that_has_untyped_template
    children:
      - pattern: has_untyped_template
        template: templates.untyped
  - id: plain_schema_child
    template: templates.example
`
  );

  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "Template text",
    fname: "templates.day",
    vault: vault1,
  });

  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "Template text",
    fname: "templates.example",
    vault: vault1,
  });

  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "Untyped template",
    fname: "templates.untyped",
    vault: vault1,
  });
};

export const setupSchemaPresetWithNamespaceTemplateBase: PreSetupHookFunction =
  async (opts) => {
    await setupBasic(opts);
    const { wsRoot, vaults } = opts;
    const vault1 = vaults[0];
    const vault2 = vaults[1];
    NoteTestUtilsV4.createSchema({
      fname: "journal",
      wsRoot,
      vault: vault1,
      modifier: (schema) => {
        const schemas = [
          SchemaUtils.createFromSchemaOpts({
            id: "daily",
            parent: "root",
            fname: "daily",
            children: ["journal"],
            vault: vault1,
          }),
          SchemaUtils.createFromSchemaRaw({
            id: "journal",
            children: ["year"],
            vault: vault1,
          }),
          SchemaUtils.createFromSchemaRaw({
            id: "year",
            pattern: "[0-2][0-9][0-9][0-9]",
            children: ["month"],
            vault: vault1,
          }),
          SchemaUtils.createFromSchemaRaw({
            id: "month",
            pattern: "[0-9][0-9]",
            children: ["day"],
            vault: vault1,
          }),
          SchemaUtils.createFromSchemaRaw({
            id: "day",
            pattern: "[0-9][0-9]",
            namespace: true,
            template: {
              id: "journal.template",
              type: "note",
            },
            vault: vault2,
          }),
        ];
        schemas.map((s) => {
          schema.schemas[s.id] = s;
        });
        return schema;
      },
    });
    await NoteTestUtilsV4.createNote({
      wsRoot,
      body: "Journal",
      fname: "daily",
      vault: vault2,
    });
    await NoteTestUtilsV4.createNote({
      wsRoot,
      body: "Template text",
      fname: "journal.template",
      vault: vault2,
    });
  };

export const setupSchemaPresetWithNamespaceTemplateMulti: PreSetupHookFunction =
  async (opts) => {
    return setupSchemaPresetWithNamespaceTemplateBase(opts);
  };

export const setupSchemaPresetWithNamespaceTemplate: PreSetupHookFunction =
  async (opts) => {
    const vault = opts.vaults[0];
    return setupSchemaPresetWithNamespaceTemplateBase({
      ...opts,
      vaults: [vault, vault],
    });
  };

export const setupEmpty: PreSetupHookFunction = async ({ vaults, wsRoot }) => {
  const vault = vaults[0];
  await NOTE_PRESETS_V4.NOTE_EMPTY.create({
    vault,
    wsRoot,
  });
};

export const setupLinks: PreSetupHookFunction = async ({ vaults, wsRoot }) => {
  return setupLinksBase({ wsRoot, vaults: [vaults[0], vaults[0]] });
};

export const setupLinksMulti: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  return setupLinksBase({ wsRoot, vaults });
};

export const setupLinksBase: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  const vault1 = vaults[0];
  const vault2 = vaults[1];
  // create note with wikilink
  await NOTE_PRESETS_V4.NOTE_WITH_TARGET.create({
    vault: vault1,
    wsRoot,
  });
  // create note with relative wikilink
  await NOTE_PRESETS_V4.NOTE_WITH_ANCHOR_LINK.create({
    vault: vault2,
    wsRoot,
  });
  // create note with labeld wikilink
  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "[[some label|beta]]",
    fname: "omega",
    vault: vault1,
  });
};

/** Creates 2 notes with same fname in 2 different vaults, and a note named
 * "test" in second vault with both valid and invalid wikilinks.
 *
 * See [[scratch.2021.07.15.205433.inconsistent-ref-and-link-behavior]] for the
 * invalid behaviors that this is intended to test for. The only difference is
 * that vaultThree is the real vault and vault3 is the bad one.
 *
 * @returns the test note with the wikilinks.
 */
export const setupMultiVaultSameFname: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}): Promise<NoteProps> => {
  await NoteTestUtilsV4.createNote({
    fname: "eggs",
    vault: vaults[0],
    body: "vault 0",
    wsRoot,
    props: { id: "eggs-vault-0" },
  });
  await NoteTestUtilsV4.createNote({
    fname: "eggs",
    vault: vaults[1],
    body: "vault 1",
    wsRoot,
    props: { id: "eggs-vault-1" },
  });
  return NoteTestUtilsV4.createNote({
    fname: "test",
    vault: vaults[1],
    body: [
      "[[eggs]]", // 7
      "[[dendron://vault1/eggs]]", // 8
      "[[dendron://vault2/eggs]]", // 9
      "[[dendron://vaultThree/eggs]]", // 10
      "[[dendron://vault3/eggs]]", // 11
      "",
      "the test note",
    ].join("\n"),
    wsRoot,
  });
};

export const setupLinksWithVaultBase: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  await setupLinksBase({ vaults, wsRoot });
};

export const setupRefs: PreSetupHookFunction = async ({ vaults, wsRoot }) => {
  const vault = vaults[0];
  // create note with note reference
  await NOTE_PRESETS_V4.NOTE_WITH_NOTE_REF_SIMPLE.create({
    vault,
    wsRoot,
  });
  await NOTE_PRESETS_V4.NOTE_WITH_NOTE_REF_SIMPLE_TARGET.create({
    vault,
    wsRoot,
  });
  // create note with block reference
  await NOTE_PRESETS_V4.NOTE_WITH_BLOCK_REF_SIMPLE.create({
    vault,
    wsRoot,
  });
  // create note with block range reference
  await NOTE_PRESETS_V4.NOTE_WITH_BLOCK_RANGE_REF_SIMPLE.create({
    vault,
    wsRoot,
  });
  // create note with reference offset
  await NOTE_PRESETS_V4.NOTE_WITH_REF_OFFSET.create({
    vault,
    wsRoot,
  });
  // create note with wildcard child reference
  await NOTE_PRESETS_V4.NOTE_WITH_WILDCARD_CHILD_REF.create({
    vault,
    wsRoot,
  });
  // create note with wildcard header reference
  await NOTE_PRESETS_V4.NOTE_WITH_WILDCARD_HEADER_REF.create({
    vault,
    wsRoot,
  });
  // create note with complex wildcard reference
  await NOTE_PRESETS_V4.NOTE_WITH_WILDCARD_COMPLEX.create({
    vault,
    wsRoot,
  });
  // create note with fm variables
  await NOTE_PRESETS_V4.NOTE_WITH_FM_VARIABLES.create({
    vault,
    wsRoot,
  });
};

export const ENGINE_HOOKS_BASE = {
  WITH_LINKS: setupLinksBase,
};

export const ENGINE_HOOKS = {
  setupBasic,
  setupHierarchyForLookupTests,
  setupSchemaPreseet,
  setupSchemaWithDiamondGrandchildren,
  setupSchemaWithIncludeOfDiamond,
  setupSchemaWithDiamondAndParentNamespace,
  setupSchemaPresetWithNamespaceTemplate,
  setupInlineSchema,
  setupSchemaWithExpansion,
  setupNoteRefRecursive,
  setupJournals,
  setupEmpty,
  setupLinks,
  setupRefs,
};

export const ENGINE_HOOKS_MULTI = {
  setupBasicMulti,
  setupLinksMulti,
  setupSchemaPresetWithNamespaceTemplateMulti,
  setupMultiVaultSameFname,
};
