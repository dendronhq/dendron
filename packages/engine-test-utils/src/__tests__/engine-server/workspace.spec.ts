import { DVault, NoteProps } from "@dendronhq/common-all";
import { tmpDir } from "@dendronhq/common-server";
import { NoteTestUtilsV4 } from "@dendronhq/common-test-utils";
import {
  DConfig,
  SeedService,
  SyncActionStatus,
  WorkspaceService,
  WorkspaceUtils,
} from "@dendronhq/engine-server";
import fs from "fs-extra";
import path from "path";
import { TestConfigUtils } from "../../config";
import {
  runEngineTestV5,
  setupWS,
  TestEngineUtils,
  testWithEngine,
} from "../../engine";
import { ENGINE_HOOKS } from "../../presets";
import {
  checkDir,
  checkFile,
  checkNotInDir,
  checkVaults,
  GitTestUtils,
} from "../../utils";
import { TestSeedUtils } from "../../utils/seed";

describe("WorkspaceUtils", () => {
  describe("findWSRoot", () => {
    it("ok: same path", async () => {
      await runEngineTestV5(
        async ({ wsRoot }) => {
          process.chdir(wsRoot);
          const wsPath = WorkspaceUtils.findWSRoot();
          // realPathSync necessary because of symlinks created by tmpdir
          // see https://stackoverflow.com/questions/31843087/unexpectedly-in-private-folder-when-changing-the-working-directory-to-a-temp-f
          expect(wsPath).toEqual(fs.realpathSync(wsRoot));
        },
        {
          expect,
        }
      );
    });

    it("ok: a few directories down", async () => {
      await runEngineTestV5(
        async ({ wsRoot }) => {
          const cwd = path.join(wsRoot, "one", "two");
          fs.ensureDirSync(cwd);
          process.chdir(cwd);
          const wsPath = WorkspaceUtils.findWSRoot();
          expect(wsPath).toEqual(fs.realpathSync(wsRoot));
        },
        {
          expect,
        }
      );
    });

    it("fail: exceed 3 directroies down", async () => {
      await runEngineTestV5(
        async ({ wsRoot }) => {
          const cwd = path.join(wsRoot, "one", "two", "three", "four");
          fs.ensureDirSync(cwd);
          process.chdir(cwd);
          const wsPath = WorkspaceUtils.findWSRoot();
          expect(wsPath).toBeUndefined();
        },
        {
          expect,
        }
      );
    });
  });
});

describe("WorkspaceService", () => {
  describe("create", () => {
    let homeDirStub: sinon.SinonStub;

    beforeEach(() => {
      homeDirStub = TestEngineUtils.mockHomeDir();
    });
    afterEach(() => {
      homeDirStub.restore();
    });
    test("basic", async () => {
      const wsRoot = tmpDir().name;
      const vaults = [{ fsPath: "vault1" }];
      await WorkspaceService.createWorkspace({ wsRoot, vaults });
      const gitignore = path.join(wsRoot, ".gitignore");
      expect(
        fs.readFileSync(gitignore, { encoding: "utf8" })
      ).toMatchSnapshot();
    });

    test("workspace Vault", async () => {
      const wsRoot = tmpDir().name;
      const vaults: DVault[] = [{ fsPath: "vault1", workspace: "foo" }];
      await WorkspaceService.createWorkspace({ wsRoot, vaults });
      expect(fs.existsSync(path.join(wsRoot, "foo", "vault1"))).toBeTruthy();
    });
  });

  describe("initialize", () => {
    testWithEngine("remoteVaults present", async ({ wsRoot, engine }) => {
      const root = tmpDir().name;
      await GitTestUtils.createRepoWithReadme(root);

      engine.config.vaults.push({
        fsPath: "remoteVault",
        remote: {
          type: "git",
          url: root,
        },
      });
      DConfig.writeConfig({ wsRoot, config: engine.config });

      const ws = new WorkspaceService({ wsRoot });
      const didClone = await ws.initialize({
        onSyncVaultsProgress: () => {},
        onSyncVaultsEnd: () => {},
      });
      expect(didClone).toEqual(true);
      expect(
        fs.existsSync(path.join(wsRoot, "remoteVault", "README.md"))
      ).toBeTruthy();
    });

    test("remote workspace present", async () => {
      await runEngineTestV5(
        async ({ wsRoot, vaults }) => {
          const { wsRoot: remoteDir } = await setupWS({
            vaults: [{ fsPath: "vault1" }],
            asRemote: true,
          });
          TestConfigUtils.withConfig(
            (config) => {
              config.workspaces = {
                remoteVault: {
                  remote: {
                    type: "git",
                    url: remoteDir,
                  },
                },
              };
              return config;
            },
            { wsRoot }
          );
          await new WorkspaceService({ wsRoot }).addVault({
            vault: { fsPath: "vault1", workspace: "remoteVault" },
            updateWorkspace: true,
            updateConfig: true,
          });
          const ws = new WorkspaceService({ wsRoot });
          const didClone = await ws.initialize({
            onSyncVaultsProgress: () => {},
            onSyncVaultsEnd: () => {},
          });
          expect(didClone).toEqual(true);
          expect(
            fs.existsSync(path.join(wsRoot, "remoteVault", "vault1", "root.md"))
          ).toBeTruthy();
          checkVaults(
            {
              wsRoot,
              vaults: [
                {
                  fsPath: "vault1",
                  workspace: "remoteVault",
                } as DVault,
              ].concat(vaults),
            },
            expect
          );
        },
        {
          expect,
          preSetupHook: async (opts) => {
            await ENGINE_HOOKS.setupBasic(opts);
          },
          addVSWorkspace: true,
        }
      );
    });

    test("remote seed present", async () => {
      await runEngineTestV5(
        async ({ engine, wsRoot, vaults }) => {
          const tmp = tmpDir().name;
          const { registryFile, seedDict } =
            await TestSeedUtils.createSeedRegistry({
              engine,
              wsRoot: tmp,
            });
          const id = TestSeedUtils.defaultSeedId();
          const seedService = new SeedService({ wsRoot, registryFile });
          const seed = seedDict[TestSeedUtils.defaultSeedId()];
          await seedService.addSeedMetadata({ seed, wsRoot });
          const wsService = new WorkspaceService({ wsRoot, seedService });
          const didClone = await wsService.initialize({
            onSyncVaultsProgress: () => {},
            onSyncVaultsEnd: () => {},
          });
          expect(didClone).toEqual(true);

          // seed should be added
          await checkDir(
            { fpath: path.join(wsRoot, "seeds", id), snapshot: true },
            "dendron.yml",
            "seed.yml",
            "vault"
          );
          // seed.yml should not be present in the workspace
          await checkNotInDir(
            {
              fpath: wsRoot,
              snapshot: true,
            },
            "seed.yml"
          );
          await checkFile({
            fpath: path.join(wsRoot, "dendron.yml"),
          });
          await checkFile(
            {
              fpath: path.join(wsRoot, "dendron.code-workspace"),
              snapshot: false,
            },
            // Necessary for windows test-compat:
            path.join(`seeds`, `${id}`, `vault`).replace(/\\/g, "\\\\")
          );
          checkVaults(
            {
              wsRoot,
              vaults: [
                { fsPath: "vault", seed: id, name: id } as DVault,
              ].concat(vaults),
            },
            expect
          );
        },
        {
          expect,
          addVSWorkspace: true,
        }
      );
    });

    // testWithEngine(
    //   "remoteVaults present as workspace vault",
    //   async ({ wsRoot, engine }) => {
    //     const {wsRoot: wsRootRemote} = await runEngineTestV5(async ()=>{}, {
    //       vaults: [{
    //         fsPath: "gamma"
    //       }],
    //       expect, setupOnly: true, initGit: true})

    //     engine.config.vaults.push({
    //       fsPath: "repos/gamma",
    //       remote: {
    //         type: "git",
    //         url: wsRootRemote,
    //       },
    //     });
    //     DConfig.writeConfig({ wsRoot, config: engine.config });
    //     expect(engine.config).toMatchSnapshot();
    //     const ws = new WorkspaceService({ wsRoot });
    //     const didClone = await ws.initialize({
    //       onSyncVaultsProgress: () => { },
    //       onSyncVaultsEnd: () => { },
    //     });
    //     expect(didClone).toEqual(true);
    //     expect(
    //       fs.existsSync(path.join(wsRoot, "repos", "gamma", "dendron.yml"))
    //     ).toBeTruthy();
    //   }, {only: true }
    // );

    testWithEngine(
      "remoteVaults present, no initializeRemoteVaults",
      async ({ wsRoot, engine }) => {
        const root = tmpDir().name;
        await GitTestUtils.createRepoWithReadme(root);

        engine.config.vaults.push({
          fsPath: "remoteVault",
          remote: {
            type: "git",
            url: root,
          },
        });
        engine.config.initializeRemoteVaults = false;
        DConfig.writeConfig({ wsRoot, config: engine.config });
        const ws = new WorkspaceService({ wsRoot });
        const didClone = await ws.initialize({
          onSyncVaultsProgress: () => {},
          onSyncVaultsEnd: () => {},
        });
        expect(didClone).toBeFalsy();
        expect(
          fs.existsSync(path.join(wsRoot, "remoteVault", "README.md"))
        ).toBeFalsy();
      }
    );
  });

  testWithEngine(
    "commitAll",
    async ({ wsRoot, vaults }) => {
      await NoteTestUtilsV4.modifyNoteByPath(
        { wsRoot, vault: vaults[0], fname: "foo" },
        (note: NoteProps) => {
          note.body += "\n Foo";
          return note;
        }
      );
      await NoteTestUtilsV4.createNote({
        fname: "bar",
        vault: vaults[1],
        wsRoot,
      });
      const resp = await new WorkspaceService({
        wsRoot,
      }).commitAndAddAll();
      expect(resp.length).toEqual(3);
      expect(
        resp.filter((r) => r.status === SyncActionStatus.DONE).length
      ).toEqual(2);
    },
    { initGit: true }
  );
});
