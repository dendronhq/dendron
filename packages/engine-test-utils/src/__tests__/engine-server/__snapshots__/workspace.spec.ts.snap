// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkspaceService create basic 1`] = `
"node_modules
.dendron.*
build

"
`;

exports[`WorkspaceService initialize remote seed present 1`] = `".git .gitignore dendron.code-workspace dendron.yml seed.yml vault"`;

exports[`WorkspaceService initialize remote seed present 2`] = `"dendron.code-workspace dendron.yml hooks seeds vault1 vault2 vault3"`;